frontends/concrete-python/concrete/fhe/mlir/processors/assign_bit_widths.py

```diff
@@ -13,266 +13,346 @@
 
 class AssignBitWidths(GraphProcessor):
     """
-    Assign a precision to all nodes inputs/output.
-
-    The precisions are compatible graph constraints and MLIR.
-    There are two modes:
-        - single precision: where all encrypted values have the same precision.
-        - multi precision: where encrypted values can have different precisions.
+    AssignBitWidths graph processor, to assign proper bit-widths to be compatible with FHE.
     """
 
     def __init__(self, single_precision=False):
         self.single_precision = single_precision
 
     def apply(self, graph: Graph):
-        nodes = graph.query_nodes()
-        for node in nodes:
-            assert isinstance(node.output.dtype, Integer)
-            node.properties["original_bit_width"] = node.output.dtype.bit_width
+        solver = z3.Solver()
 
-        if self.single_precision:
-            assign_single_precision(nodes)
-        else:
-            assign_multi_precision(graph, nodes)
+        bit_widths: Dict[Node, z3.Int] = {}
 
+        additional_constraints = AdditionalConstraints(solver, graph, bit_widths)
 
-def assign_single_precision(nodes: list[Node]):
-    """Assign one single encryption precision to all nodes."""
-    p = required_encrypted_bitwidth(nodes)
-    for node in nodes:
-        assign_precisions_1_node(node, p, p)
+        nodes = graph.query_nodes(ordered=True)
+        for i, node in enumerate(nodes):
+            bit_width = z3.Int(f"%{i}")
+            bit_widths[node] = bit_width
 
+            assert isinstance(node.output.dtype, Integer)
+            solver.add(bit_width >= node.output.dtype.bit_width)
 
-def assign_precisions_1_node(node: Node, output_p: int, inputs_p: int):
-    """Assign input/output precision to a single node.
+            additional_constraints.generate_for(node, bit_width)
 
-    Precision are adjusted to match different use, e.g. encrypted and constant case.
-    """
-    assert isinstance(node.output.dtype, Integer)
-    if node.output.is_encrypted:
-        node.output.dtype.bit_width = output_p
-    else:
-        node.output.dtype.bit_width = output_p + 1
+        if self.single_precision:
```

Comment:
is equality always possible ? can't it be contradicted by a general rule ?

Reply:
It is always possible yes, imagine in FHE, we can set every bit-width to the same, but a big one :)